src/socket_poller.cpp

```diff
@@ -401,6 +413,97 @@ void zmq::socket_poller_t::rebuild ()
     need_rebuild = false;
 }
 
+void zmq::socket_poller_t::zero_trail_events(
+                                         zmq::socket_poller_t::event_t *events_,
+                                         int n_events_,
+                                         int found) {
+    for (int i = found; i < n_events_; ++i) {
+        events_[i].socket = NULL;
+        events_[i].fd = 0;
+        events_[i].user_data = NULL;
+        events_[i].events = 0;
+        events_[i].zid = NULL;
+        events_[i].zid_len = 0;
+    }
+}
+
+int zmq::socket_poller_t::check_events(zmq::socket_poller_t::event_t *events_,
+                                                                 int n_events_,
+                                                                 int& found) {
+        for (items_t::iterator it = items.begin (); it != items.end () &&
+                                                      found < n_events_; ++it) {
+
+            events_[found].socket = NULL;
+            events_[found].fd = 0;
+            events_[found].user_data = NULL;
+            events_[found].events = 0;
+
+            //  The poll item is a 0MQ socket. Retrieve pending events
+            //  using the ZMQ_EVENTS socket option.
+            if (it->socket) {
+                uint32_t events;
+                size_t events_size;
+
+                if (it->zid_len == 0){
+                    //  Regular poller item
+                    events_size = sizeof (uint32_t);
+
+                    if (it->socket->getsockopt (ZMQ_EVENTS, &events,
+                                                            &events_size) == -1)
+                        return -1;
+                } else {
+                    //  Peer poller item
+                    zmq_gso_peer_events_t peer_events;
+                    peer_events.zid = it->zid;
+                    peer_events.zid_len = it->zid_len;
+                    events_size = ZMQ_GSO_PEER_EVS_SIZE;
+
+                    if (it->socket->getsockopt (ZMQ_PEER_EVENTS,
+						(void*)&peer_events,
+                                                &events_size) == -1)
+                        return -1;
+                    events = peer_events.events;
+                }
+
+                if (it->events & events) {
+                    events_[found].socket = it->socket;
+                    events_[found].user_data = it->user_data;
+                    events_[found].events = it->events & events;
+                    events_[found].zid = it->zid;
+                    events_[found].zid_len = it->zid_len;
+                    ++found;
+                }
+            }
+            //  Else, the poll item is a raw file descriptor, simply convert
+            //  the events to zmq_pollitem_t-style format.
+            else {
+                short revents = pollfds [it->pollfd_index].revents;
```

Comment:
what about windows case?

Reply:
Well that code is a straight porting of the two repeated blobs of code:

https://github.com/zeromq/libzmq/blob/master/src/socket_poller.cpp#L468-L514
https://github.com/zeromq/libzmq/blob/master/src/socket_poller.cpp#L633-L670

If anything is missing for Windows, then it's also missing in master. But I don't think so, right?

Reply:
the code is not exactly the same... take a closer look

Reply:
the handling of FD case

Reply:
@somdoron I will have a closer look when doing the refactor PR first. Thanks